Better Idle performance

Gusted opened the pull request:
- Uses a prototype override to detect changes in the style.
- This will improve idle CPU consumption of the browser as it won't have to run `requestAnimationFrame` to detect changes.
- Most of the understanding of prototype overriding came from @tophf's [XRay-Styler](https://github.com/tophf/XRayStyler/). Credits to him!
- NO WINDOW POLLUTING!!
- Resolves #2070
- Resolves #1125
- Resolves #3905

Gusted commented on src/inject/dynamic-theme/style-manager.ts:
```diff
@@ -318,44 +318,67 @@ export function manageStyle(element: StyleElement, {update, loadingStart, loadin
+                function proxyAddRule(selector, style, index) {
+                    addRuleFunction.value.call(this, selector, style, index);
+                    this.ownerNode.dispatchEvent(event);
+                    // Check why it returns -1 https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/addRule#Return_value.
```
https://drafts.csswg.org/cssom/#dom-cssstylesheet-deleterule check definition 8.
https://searchfox.org/mozilla-central/source/layout/style/StyleSheet.cpp#640 Firefox return 1;
https://github.com/chromium/chromium/blob/master/third_party/blink/renderer/core/css/css_style_sheet.cc#L442 Chromium return 1;

Gusted commented on src/inject/dynamic-theme/style-manager.ts:
```diff
@@ -318,44 +318,67 @@ export function manageStyle(element: StyleElement, {update, loadingStart, loadin
+                    return -1;
+                }
+                function proxyInsertRule(rule, index) {
+                    // Cache the return value of insertRule as insertRule has the chance to not return the originial index.
```
https://github.com/chromium/chromium/blob/master/third_party/blink/renderer/core/css/css_style_sheet.cc#L322-L367 can return `0` due to errors.
https://searchfox.org/mozilla-central/source/layout/style/StyleSheet.cpp#588-597 can return `0` due to errors.

Gusted commented:
[darkreader.zip](https://github.com/darkreader/darkreader/files/5470038/darkreader.zip)
Debug copies of this Branch.

And a test page it should make the background-color `rgb(89, 0, 16)`
[aaa.zip](https://github.com/darkreader/darkreader/files/5470039/aaa.zip)

ddnexus commented:
Just tried the debug copy on the same system that was consuming quite a lot on idle, and it works as expected now. 
Great job! :1st_place_medal:

Gusted commented:
> 
> 
> Just tried the debug copy on the same system that was consuming quite a lot on idle, and it works as expected now.
> Great job! 🥇

Great to hear that!

Gusted commented:
Cleaned up a lot of code and made it less messy and more understandable also fixed some bugs. https://github.com/darkreader/darkreader/pull/4005/commits/9e7b381b5efef2f9e85ccde6745731b9bd8e5450

Improved the clean up of the prototype override. https://github.com/darkreader/darkreader/pull/4005/commits/1044344498d9d836d0aa73339634386f9b4b6754

here the build binaries for Firefox & Chromium.
[darkreader.zip](https://github.com/darkreader/darkreader/files/5471852/darkreader.zip)

It should be stable enough now.

ddnexus commented:
I retried both builds, the difference between dynamic and non-dynamic mode on idle is almost 0 almost on all pages that I tried.

However that doesn't seem to apply to the gmail inbox page. In that case the improvement is just marginal (i.e. CPU difference on idle is still about 10% more for dynamic).

Maybe that's worth some more analysis, because it could be a more general case not limited to gmail.

Here is the profile:
[Profile-20201102T105255.zip](https://github.com/darkreader/darkreader/files/5472731/Profile-20201102T105255.zip)

Gusted commented:
> However that doesn't seem to apply to the gmail inbox page. In that case the improvement is just marginal (i.e. CPU difference on idle is still about 10% more for dynamic).
> 
> Maybe that's worth some more analysis, because it could be a more general case not limited to gmail.
> 
> Here is the profile:
> [Profile-20201102T105255.zip](https://github.com/darkreader/darkreader/files/5472731/Profile-20201102T105255.zip)

The faulty code is https://github.com/darkreader/darkreader/blob/master/src/inject/dynamic-theme/watch.ts#L47

ddnexus commented:
Good that you already know the cause, but is there also a solution?

Gusted commented:
> Good that you already know the cause, but is there also a solution?

I have a idea for a solution that requires a rewrite of this `defined` logic so expect a new commit to fix this within a few days.

alexanderby commented on src/inject/dynamic-theme/index.ts:
```diff
@@ -106,16 +109,28 @@ function createStaticStyleOverrides() {
+    setupNodePositionWatcher(variableStyle, 'variables');
+
+    const proxyScript = createOrUpdateElement('darkreader--proxy', undefined, 'script');
+    proxyScript.textContent = `// First get the originial 'native code' of the function.
```
I suggest you to move the whole function into a separate file. Avoid polluting window scope.
E.g. use something like
```javascript
import {injectProxy} from './stylesheet-proxy';
proxyScript.textContent = `(${goodFunctionName})()`;
```
It will convert the function to a string and execute.

alexanderby commented on src/inject/dynamic-theme/index.ts:
```diff
@@ -422,6 +437,18 @@ export function createOrUpdateDynamicTheme(filterConfig: FilterConfig, dynamicTh
+function removeProxy() {
+    const proxy = document.head.querySelector('.darkreader--proxy');
+    if (proxy) {
+        const Dproxy = document.createElement('script');
```
`Dproxy`? :) Try some different meaningful lowercase name.

alexanderby commented on src/inject/dynamic-theme/index.ts:
```diff
@@ -106,16 +109,28 @@ function createStaticStyleOverrides() {
+        ${removeProxyOnPrototype.toString()}
+        injectProxy();`;
+    document.head.insertBefore(proxyScript, variableStyle.nextSibling);
+    setupNodePositionWatcher(proxyScript, 'proxy');
```
I don't think there's a need to watch for its position.

alexanderby commented on src/inject/dynamic-theme/index.ts:
```diff
@@ -25,50 +26,52 @@ let filter: FilterConfig = null;
-        style.classList.add('darkreader');
-        style.classList.add(className);
-        style.media = 'screen';
+function createOrUpdateElement(className: string, root: ParentNode = document.head || document, elementType = 'style') {
```
Let's create 2 functions: one for style and one for script, because it is not obvious that calling `createOrUpdateElement(className)` returns a style element.

alexanderby commented on src/inject/dynamic-theme/style-manager.ts:
```diff
@@ -318,50 +318,22 @@ export function manageStyle(element: StyleElement, {update, loadingStart, loadin
-
-    function unsubscribeFromSheetChanges() {
-        cancelAnimationFrame(rulesCheckFrameId);
+        element.addEventListener('updateSheet', update);
```
There's a small chance of events name collision. Add some `darkreader` prefix to the event name.

alexanderby commented on src/inject/dynamic-theme/stylesheet-proxy.ts:
```diff
@@ -0,0 +1,43 @@
+    function proxyAddRule(selector?: string, style?: string, index?: number): number {
+        window['addRuleFunction'].value.call(this, selector, style, index);
+        this.ownerNode.dispatchEvent(updateSheetEvent);
+        // Should always returns -1 https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/addRule#Return_value.
```
You can remove all comments except this one.

alexanderby commented on src/inject/dynamic-theme/stylesheet-proxy.ts:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    // Define the prototype of the CSSStyleSheet to use the proxy function.
+    Object.defineProperty(CSSStyleSheet.prototype, 'addRule', {...window['addRuleFunction'], value: proxyAddRule});
```
So, try doing the same without polluting the global scope.

narcolepticinsomniac commented:
Needless to say, this PR is awesome as well. Previously, Chromium devtools were practically unusable when the dynamic theme was active on a page. This resolves whatever weird conflict element highlighting has with `requestAnimationFrame`, and the non-stop, ridiculous CPU consumption that comes along with it. Also resolves some other weird bugs I've encountered on a somewhat regular basis on certain sites.

alexanderby commented on src/inject/dynamic-theme/stylesheet-proxy.ts:
```diff
@@ -0,0 +1,43 @@
+
+export function removeProxyOnPrototype() {
+    // Define the prototype of the CSSStyleSheet to the orginial function.
+    Object.defineProperty(CSSStyleSheet.prototype, 'addRule', {...window['addRuleFunction']});
```
Instead try launching the cleanup by dispatching a custom event on document, or saving a removal function into window scope like `__darkreader__removeCSSStyleSheetProxy__()` (more ugly).

alexanderby commented:
One ugly thing about adding style is that some sites can have CSP no inline style etc.

alexanderby commented on src/inject/dynamic-theme/stylesheet-proxy.ts:
```diff
@@ -0,0 +1,46 @@
+export function injectProxy() {
+    if (!window['__darkreader__addRuleFunction']) {
+        window['__darkreader__addRuleFunction'] = Object.getOwnPropertyDescriptor(CSSStyleSheet.prototype, 'addRule');
```
You can just store the descriptors in the function scope like `const addRuleDescriptor = Object.getOwnPropertyDescriptor(CSSStyleSheet.prototype, 'addRule')` :)

alexanderby reviewed (approved):
Great!

Now the new problem arises: if some style sheet is changed synchronously many times, the update will be called too often. The `throttle` function can come into rescue. Or maybe `queueMicrotask` can help there. Let it be a problem of another PR.

Gusted commented:
> Let it be a problem of another PR.

It will be a issue. I have spend too many hours on this PR as well.

Gusted commented:
> Good that you already know the cause, but is there also a solution?

#4055 It's possible however I still know half of what the defined Elements, so it's up to @alexanderby for this one.
